### DISCUSSION TITLE
[Feature] NSFW Tensorflow detection

### DISCUSSION THREAD

**Discussion by spusuf** [immich-app/immich#3833] [Feature Request]
### The feature

Using a nsfw detector to separate photos and exclude them from timeline and faces. Models for tensorflow are available such as:
https://github.com/infinitered/nsfwjs

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by AngelaDMerkel** [comment]
I think this is the best approach to #2451 because it would alleviate some of the challenges with filtering older or previously uploaded images. PhotoPrism does something similar and the relevant script is [here](https://github.com/photoprism/photoprism/blob/develop/scripts/download-nsfw.sh). I cannot comment on the accuracy, but this seems like a very reasonable approach. Training a model to recognise NSFW content would seem like a lot of work on the developers part, and creating a useful dataset to train on would also seem to be very difficult and very unlikely.

**Comment by blistovmhz** [comment]
Fully agree, but the dataset and trained model are already available. this is really just a drop in plugin at this point. I hope someone gets on it as I just don't have the time currently.
https://github.com/MaybeShewill-CV/nsfw-classification-tensorflow

**Reply by aviv926** [reply]
I think that using clip's Embeddings that are created anyway would be a much more accurate and resource efficient idea
There is already something ready like that, you just need to know how to implement it

https://github.com/LAION-AI/CLIP-based-NSFW-Detector

**Comment by radh21301** [comment]
I think could be a huge QOL feature. Imagine all the times you wanted to show a friend the recent pics you took, and voila, there is this ridiculous topless picture you took to track your gym life.

**Solution:** I think using the existing CLIP algorithm works wonders. I just searched for 'Intimate', 'shirtless' -> and the results were amazing. And using an existing algorithm would make it easy to implement and maintain as well.

The user settings page can let us list of comma separated 'smart search terms' that should be filtered out. And the top of the GUI, has an icon that can be toggled on/off to toggle this mode. Something like this:
<img width="612" alt="image" src="https://github.com/user-attachments/assets/e70df54b-b2af-4cb2-9418-78878ad03b12">


EDIT: This would not work. CLIP results are not binary. It does not say if an image has a car OR no car. It tells us the chance that an image has a car (something like probability). Probability is never binary. Even an image with a building will be assigned a score (it might be close to 0, but a score all the same). And Immich results in the search page is infinite- as you keep scrolling down, the lesser the chance that the image matches your search term.
So my suggestion earlier would hide all images in your library - unless immich uses a cutoff score. Better to use the clip based nsfw models I suppose.

**Comment by bo0tzz** [comment]
This should become possible with #1673

**Reply by aviv926** [reply]
Would it be possible to add a CLIP model [designed for this](https://github.com/immich-app/immich/discussions/3833#discussioncomment-7667402)? Because smart albums that rely on a CLIP model that has not been trained for it can sometimes miss potential photos
